react/components/Modal/index.tsx

```diff
@@ -0,0 +1,265 @@
+  showCloseIcon?: boolean
+  bottomBar?: React.ReactNode
+  title?: React.ReactNode
+  closeOnEsc?: boolean
+  showBottomBarBorder?: boolean
+  onCloseTransitionFinish?: () => unknown
+  centered?: boolean
+  size?: 'small' | 'medium' | 'large'
+  responsiveFullScreen?: boolean
+}
+
+type OverlayProps = Required<
+  Pick<Props, 'isOpen' | 'centered' | 'closeOnOverlayClick' | 'onClose'>
+> &
+  Pick<Props, 'onCloseTransitionFinish' | 'container'>
+
+type ContentProps = Required<
+  Pick<
+    Props,
+    | 'onClose'
+    | 'showCloseIcon'
+    | 'showTopBar'
+    | 'children'
+    | 'showBottomBarBorder'
+    | 'responsiveFullScreen'
+    | 'size'
+    | 'closeOnEsc'
+  >
+> &
+  Pick<Props, 'title' | 'bottomBar'> &
+  Omit<React.HTMLAttributes<HTMLDivElement>, 'title'>
+
+export const ModalOverlay: FC<OverlayProps> = ({
+  isOpen,
+  onClose,
+  centered,
+  container,
+  closeOnOverlayClick,
+  onCloseTransitionFinish,
+  children,
+}) => {
+  const [showPortal, setShowPortal] = useState<boolean>(isOpen)
+  const [focusReturnNode, setFocusReturnNode] = useState<
+    HTMLElement | Element | null
+  >()
+
+  useEnhancedEffect(() => {
+    if (isOpen) {
+      setFocusReturnNode(document.activeElement)
+    }
+
+    return () => {
+      if (
+        focusReturnNode instanceof HTMLElement &&
+        document.contains(focusReturnNode)
+      ) {
+        focusReturnNode?.focus()
+      }
+    }
+  })
+
+  useEnhancedEffect(() => {
+    if (!isOpen) {
+      return
+    }
+    setShowPortal(isOpen)
+    if (canUseDOM) document.body.classList.add(styles.hiddenScroll)
+  }, [isOpen])
+
+  const handleClick = () => {
+    if (!closeOnOverlayClick) return
+    onClose()
+  }
+
+  const handleAnimationEnd = () => {
+    if (isOpen) return
+    onCloseTransitionFinish?.()
+    setShowPortal(false)
+    if (canUseDOM) document.body.classList.remove(styles.hiddenScroll)
+  }
+
+  useEffect(() => {
+    return () => {
+      if (canUseDOM) document.body.classList.remove(styles.hiddenScroll)
+    }
+  }, [])
+
+  return showPortal
+    ? createPortal(
+        /** This rule can be disabled because we are not using the onClick property to click
+         * the element itself but to capture outside clicks */
+        // eslint-disable-next-line jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events
+        <div
+          className={classNames(
+            'vtex.modal__overlay flex fixed z-max overflow-hidden bg-black-70 absolute--fill outline-0',
+            isOpen ? styles.openAnimation : styles.closeAnimation,
+            {
+              'items-start': !centered,
+              'items-center': centered,
+            }
+          )}
+          tabIndex={-1}
+          onClick={handleClick}
+          onAnimationEnd={handleAnimationEnd}
+          data-testid="modal__overlay"
+          role="presentation"
+        >
+          <FocusTrap>{children}</FocusTrap>
+        </div>,
+        container ?? document.body
```

Comment:
When `canUseDOM` is false the modal won't open because `createPortal` doesn't work :worried: it's a use case that is also happening in the current modal and I'm still not sure how to fix it